packages/flutter_tools/lib/src/vmservice.dart

```diff
@@ -332,28 +347,60 @@ class VMService {
     Device device,
   }) async {
     final Uri wsUri = httpUri.replace(scheme: 'ws', path: globals.fs.path.join(httpUri.path, 'ws'));
-    final StreamChannel<String> channel = await _openChannel(wsUri, compression: compression);
-    final rpc.Peer peer = rpc.Peer.withoutJson(jsonDocument.bind(channel), onUnhandledError: _unhandledError);
+    final io.WebSocket channel = await _openChannel(wsUri.toString(), compression: compression);
+    final StreamController<dynamic> primary = StreamController<dynamic>();
+    final StreamController<dynamic> secondary = StreamController<dynamic>();
+
+    channel.listen((dynamic data) {
+      primary.add(data);
+      secondary.add(data);
+    }, onDone: ()  {
+      primary.close();
+      secondary.close();
+    }, onError: (dynamic error, StackTrace stackTrace) {
+      primary.addError(error, stackTrace);
+      secondary.addError(error, stackTrace);
+    });
+
+    // Create an instance of the package:vm_service API in addition to the flutter
+    // tool's to allow gradual migration.
+    final Completer<void> streamClosedCompleter = Completer<void>();
+
+    final vm_service.VmService delegateService = vm_service.VmService(
+      primary.stream,
+      channel.add,
+      log: null,
+      disposeHandler: () async {
+        if (!streamClosedCompleter.isCompleted) {
```

Comment:
Missing check for is completed. The old VM service might have been OK with being torn down multiple times though, not sure. At any rate I can't repro the test failure anymore

Reply:
If we're trying to tear it down twice, could we also be trying to use it after closing it in some other way?

Reply:
I've looked into it a bit more and I'm not certain - I can't consistently repro the crash. I think the stream logic is just too complicated right now, and should be a bit simpler once I no longer need to duplicate events